hubspot3/tickets.py

```diff
@@ -78,3 +90,171 @@ def get_all(self, limit: int = -1, **options) -> list:
+        else:
+            params = {}
+        batch = self._call(
+            "objects/tickets/batch-read",
+            method="POST",
+            doseq=True,
+            params=params,
+            data={'ids': ids}
+        )
+        # Returning a dict with IDs as keys
+        return batch
+
+    def get_batch(self, ids, extra_properties: Union[list, str] = None):
+        """given a batch of ticket, get more of their info"""
+        batch = self._get_batch(ids, extra_properties=extra_properties, with_history=False)
+        return [prettify(batch[ticket], id_key="objectId") for ticket in batch]
+
+    def get_batch_with_history(self, ids, extra_properties: Union[list, str] = None):
+        """given a batch of ticket, get more of their info with history"""
+        batch = self._get_batch(ids, extra_properties=extra_properties, with_history=True)
+        return batch
+
+    def _get_recent(
+        self,
+        recency_type: str,
+        limit: int = -1,
+        vid_offset: int = 0,
+        time_offset: int = 0,
+        max_time_diff_ms: int = 60000,
+        # Correct default value waiting for
+        # https://community.hubspot.com/t5/APIs-Integrations/Time-difference-between-ticket-s-quot-get-log-of-changes-quot/m-p/297909#M27982
+        **options
+    ):
+
+        finished = False
+        query_limit = 100  # max according to the docs
+        limited = limit > 0
+        if limited and limit < query_limit:
+            query_limit = limit
+        total_tickets = 0
+        while not finished:
+            params = {"count": query_limit}
+            if vid_offset:
+                params["objectId"] = vid_offset
+            if time_offset:
+                params["timestamp"] = int(time_offset)
+            params['changeType'] = recency_type
+            batch = self._call(
+                "change-log/tickets",
+                method="GET",
+                params=params,
+                doseq=True,
+                **options
+            )
+
+            total_tickets += len(batch)
+            finished = (len(batch) == 0) or (limited and total_tickets >= limit)
+            if len(batch) > 0:
+                vid_offset = batch[-1]['objectId']
+                time_offset = batch[-1]['timestamp']
+                changes = self._parse_changes(batch, query_limit, max_time_diff_ms)
+                yield changes
+
+    def _parse_changes(self, batch, query_limit, max_time_diff_ms):
+        # First lets group changes by deal id
+        changes_by_id = [(ticket_change['objectId'], ticket_change) for ticket_change in batch]
+        changes_by_id = sorted(changes_by_id, key=itemgetter(0))
+        changes_by_id = dict((key, list(value)) for key, value in
+                             itertools.groupby(changes_by_id, key=itemgetter(0)))
+        # Loop through ids in batches of 100 (limit of the batch API)
+        for index in range(int(len(changes_by_id) / query_limit) + 1):
```

Comment:
There's a better solution for `window` functions, check it out: https://stackoverflow.com/questions/6822725/rolling-or-sliding-window-iterator

Reply:
removed this windowed in the refactoring